### PR TITLE
feat(argo-cd): manage CRD by Helm

### PR DESCRIPTION
**Pull request by yu-croco**
This PR resolves https://github.com/argoproj/argo-helm/issues/1336 and https://github.com/argoproj/argo-helm/issues/1259 . 🙋 

---

Signed-off-by: yu-croco <yuki.kita22@gmail.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] I have bumped the chart version according to [versioning](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#versioning)
* [x] I have updated the documentation according to [documentation](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#documentation)
* [x] I have updated the chart changelog with all the changes that come with this pull request according to [changelog](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#changelog).
* [x] Any new values are backwards compatible and/or have sensible default.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md).
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/stable/developer-guide/ci/)).

Changes are automatically published when merged to `main`. They are not published on branches.

### CODE REVIEW

File: charts/argo-cd/values.yaml [hunk 1]
```diff
@@ -9,6 +9,11 @@ fullnameOverride: ""
 # -- Override the Kubernetes version, which is used to evaluate certain manifests
 kubeVersionOverride: ""
 
+## Custom resource configuration
+crds:
+  # -- Install and upgrade CRDs
+  install: true
```
Comment (yu-croco) [inline charts/argo-cd/values.yaml:15]: This is same as [argo-events](https://github.com/argoproj/argo-helm/blob/main/charts/argo-events/values.yaml#L18-L21).

File: charts/argo-cd/README.md.gotmpl [hunk 2]
```diff
@@ -84,15 +84,25 @@ Changes in the `CustomResourceDefinition` resources shall be fixed easily by cop
...
+
+```bash
+YOUR_ARGOCD_NAMESPACE="" # e.g. argo-cd
+YOUR_ARGOCD_RELEASENAME="" # e.g. argo-cd
+
+for crd in "applications.argoproj.io" "applicationsets.argoproj.io" "argocdextensions.argoproj.io" "appprojects.argoproj.io"; do
+  kubectl label --overwrite crd $crd app.kubernetes.io/managed-by=Helm
+  kubectl annotate --overwrite crd $crd meta.helm.sh/release-namespace="$YOUR_ARGOCD_NAMESPACE"
+  kubectl annotate --overwrite crd $crd meta.helm.sh/release-name="$YOUR_ARGOCD_RELEASENAME"
+done
```
Comment (yu-croco) [inline charts/argo-cd/README.md.gotmpl:105]: same instruction as https://github.com/argoproj/argo-helm/pull/1269 📝

File: charts/argo-cd/README.md [hunk 3]
```diff
@@ -84,15 +84,34 @@ Changes in the `CustomResourceDefinition` resources shall be fixed easily by cop
...
 
-Please use `kubectl` to upgrade CRDs manually from [crds](crds/) folder or via the manifests from the upstream project repo:
+Please use `kubectl` to upgrade CRDs manually from [templates/crds](templates/crds/) folder or via the manifests from the upstream project repo:
 
 ```bash
-kubectl apply -k https://github.com/argoproj/argo-cd.git/manifests/crds?ref=<appVersion>
+kubectl apply -k https://github.com/argoproj/argo-cd/manifests/crds\?ref\=<appVersion>
+
+# Eg. version v2.4.2
+kubectl apply -k https://github.com/argoproj/argo-cd/manifests/crds\?ref\=v2.4.2
```
Comment (yu-croco) [inline charts/argo-cd/README.md:97]: This did not work, so fixed URL. 🙋

File: charts/argo-cd/Chart.yaml [hunk 4]
```diff
@@ -2,7 +2,7 @@ apiVersion: v2
 appVersion: v2.4.2
 description: A Helm chart for Argo CD, a declarative, GitOps continuous delivery tool for Kubernetes.
 name: argo-cd
-version: 4.9.7
+version: 4.10.0
```
Comment (yu-croco) [inline charts/argo-cd/Chart.yaml:5]: Is this versioning reasonable ...? 🤔

File: charts/argo-cd/README.md.gotmpl [hunk 5]
```diff
@@ -84,15 +84,34 @@ Changes in the `CustomResourceDefinition` resources shall be fixed easily by cop
...
 
-Please use `kubectl` to upgrade CRDs manually from [crds](crds/) folder or via the manifests from the upstream project repo:
+Please use `kubectl` to upgrade CRDs manually from [templates/crds](templates/crds/) folder or via the manifests from the upstream project repo:
 
 ```bash
-kubectl apply -k https://github.com/argoproj/argo-cd.git/manifests/crds?ref=<appVersion>
+kubectl apply -k https://github.com/argoproj/argo-cd/manifests/crds\?ref\=<appVersion>
+
+# Eg. version v2.4.2
+kubectl apply -k https://github.com/argoproj/argo-cd/manifests/crds\?ref\=v2.4.2
```
Comment (yu-croco) [inline charts/argo-cd/README.md.gotmpl:97]: It did not work before, so fixed. 🙋

**Review by davidkarlsen — APPROVED** [review]
lgtm, but I'll let it sit a bit for other to have opinions too

**Review by mbevc1 — APPROVED** [review]
Fine here as well, not sure why we haven't managed CRDs via chart before... @mkilchhofer thoughts?

**Review by mkilchhofer — CHANGES_REQUESTED** [review]
We cannot do this unless we deprecate and remove the options
- server.additionalApplications
- server.additionalProjects

I am on vacation right now so I cannot explain in detail via mobile phone ;)

Reasons for nok from me:
- ~helm does not support templating for files sitting inside the /crds folder~ *mobile view didn't show that the files were moved)*
- helm needs to apply CRDs first and wait until the API extension is ready before we can deploy a instance of that CRD, otherwise the release is failed. And IMHO we should not implement a feature which rely on an edge case which might work for some but not all cases and doesn't respect the architecture how helm is designed.

In my opition we should
1. remove the two options listed above
2. inform our customers/users (at least via README and a major version bump)
3. provide a dedicated chart for the removed features like argocd-apps or argocd-resources
4. and then we could talk about implementing that feature :)

**Comment by yu-croco** [comment]
Thank you for explaining!

> helm needs to apply CRDs first and wait until the API extension is ready before we can deploy a instance of that CRD, otherwise the release is failed. And IMHO we should not implement a feature which rely on an edge case which might work for some but not all cases and doesn't respect the architecture how helm is designed.

You're right. It's better to consider as a whole including the design of helm, instead of patching partially. 🤔 


> In my opition we should
> 
> 1. remove the two options listed above
> 2. inform our customers/users (at least via README and a major version bump)
> 3. provide a dedicated chart for the removed features like argocd-apps or argocd-resources
> 4. and then we could talk about implementing that feature :)
> 
> Another approach would be to externalize all CRDs which was recommended here: #550

I agree with it. 
May I create PR(s) to resolve the above if there were not any other opinions/objections? 🙋

**Comment by pdrastil** [comment]
@yu-croco - If you are getting inspiration from Argo Events please note the `keep` option that is not in your PR. If you uninstall the chart CRDs will be also removed together with all user-defined applications.

**Comment by yu-croco** [comment]
@pdrastil 
Thank you for more information! Oh that is important.  :eyes:
https://github.com/argoproj/argo-helm/blob/main/charts/argo-events/values.yaml#L22-L23

**Comment by yu-croco** [comment]
*Whichever method we adopt, we don't have to keep open this PR, so I've put it back in draft.

File: charts/argo-cd/templates/crds/crd-application.yaml [hunk 6]
```diff
@@ -1,3 +1,4 @@
+{{- if .Values.crds.install }}
 apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
```
Comment (mkilchhofer) [inline charts/argo-cd/templates/crds/crd-application.yaml:4]: Let's also implement `crds.keep` option as argo-events [does](https://github.com/argoproj/argo-helm/blob/argo-events-2.0.3/charts/argo-events/templates/crds/eventbus-crd.yml#L7-L9).

```suggestion
metadata:
  annotations:
    {{- if .Values.crds.keep }}
    "helm.sh/resource-policy": keep
    {{- end }}
```

Comment (mkilchhofer) [reply charts/argo-cd/templates/crds/crd-application.yaml:4]: Implemented it for you @yu-croco

Comment (yu-croco) [reply charts/argo-cd/templates/crds/crd-application.yaml:4]: Thank you!
I will check failed CI. 👀

**Comment by yu-croco** [comment]
👀 🤔 
https://github.com/argoproj/argo-helm/runs/8005988745?check_suite_focus=true

```
Creating namespace 'argo-cd-ot5rdosmp3'...
namespace/argo-cd-ot5rdosmp3 created
Error: INSTALLATION FAILED: rendered manifests contain a resource that already exists. Unable to continue with install: CustomResourceDefinition "applications.argoproj.io" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; annotation validation error: key "meta.helm.sh/release-name" must equal "argo-cd-ot5rdosmp3": current value is "argo-cd-8eonx6zuj5"; annotation validation error: key "meta.helm.sh/release-namespace" must equal "argo-cd-ot5rdosmp3": current value is "argo-cd-8eonx6zuj5"
========================================================================================================================
........................................................................................................................
==> Events of namespace argo-cd-ot5rdosmp3
........................................................................................................................
No resources found in argo-cd-ot5rdosmp3 namespace.
........................................................................................................................
<== Events of namespace argo-cd-ot5rdosmp3
........................................................................................................................
========================================================================================================================
Deleting release 'argo-cd-ot5rdosmp3'...
Error: uninstall: Release not loaded: argo-cd-ot5rdosmp3: release: not found
Error deleting Helm release: Error waiting for process: exit status 1
Deleting namespace 'argo-cd-ot5rdosmp3'...
namespace "argo-cd-ot5rdosmp3" deleted
Namespace 'argo-cd-ot5rdosmp3' terminated.
Error: Error installing charts: Error processing charts
------------------------------------------------------------------------------------------------------------------------
 ✖︎ argo-cd => (version: "5.1.0", path: "charts/argo-cd") > Error waiting for process: exit status 1
------------------------------------------------------------------------------------------------------------------------
Error installing charts: Error processing charts
Error: Process completed with exit code 1.
```

**Comment by mkilchhofer** [comment]
> How can I avoid this... 👀 🤔 https://github.com/argoproj/argo-helm/runs/8005988745?check_suite_focus=true
> 
> ```
> Creating namespace 'argo-cd-ot5rdosmp3'...
> namespace/argo-cd-ot5rdosmp3 created
> Error: INSTALLATION FAILED: rendered manifests contain a resource that already exists. Unable to continue with install: CustomResourceDefinition "applications.argoproj.io" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; annotation validation error: key "meta.helm.sh/release-name" must equal "argo-cd-ot5rdosmp3": current value is "argo-cd-8eonx6zuj5"; annotation validation error: key "meta.helm.sh/release-namespace" must equal "argo-cd-ot5rdosmp3": current value is "argo-cd-8eonx6zuj5"
> (..)

Fixed it like this [a24cf37](https://github.com/argoproj/argo-helm/pull/1342/commits/a24cf37516a32bcc5f80c5c0a40aa24017aaa58f)

**Comment by yu-croco** [comment]
@mkilchhofer 
Thank you for fixing! 👍 🙇 
I turned it into ready.

**Review by mkilchhofer — APPROVED** [review]
🎉

**Pull request merged**